### PR TITLE
Adding Uniform distribution to PyTorch

### PR DESCRIPTION
**Pull request by neerajprad**
The implementation is straightforward. A couple of notes:
 - In our discussions, we had discussed throwing an error if `log_prob` got a value outside the distribution's support. This returns `-inf` instead. I thought that throwing an error does not seem to be a good response for a perfectly valid query. For reference, Tensorflow return `-inf` for uniform but throws an error for Bernoulli (scoring `[1]` with `Bernoulli([0.0])` for example). We should be consistent with whatever convention we choose.
 - Modified `sample_n()` to use `torch.Size` instead of a tuple `(n,)`. The reason is that `tensor.new(tuple)` just takes the last dim, as compared to `tensor.new(torch.Size)`. Not sure if that's expected. cc. @apaszke. 

```
In [7]: torch.zeros(2, 2).new((3, 2)).size()
Out[7]: torch.Size([2])

In [8]: torch.zeros(2, 2).new(torch.Size((3, 2))).size()
Out[8]: torch.Size([3, 2])
```

### CODE REVIEW

**Comment by neerajprad** [comment]
cc. @apaszke, @fritzo, @alicanb

**Comment by apaszke** [comment]
Creating a tensor out of a tuple doesn't take the last dim (change it to sth else). It just treats it as a sequence (as if you gave it a list), and creates a 2-element tensor.

**Review by apaszke — APPROVED** [review]

**Comment by neerajprad** [comment]
> Creating a tensor out of a tuple doesn't take the last dim (change it to sth else). It just treats it as a sequence (as if you gave it a list), and creates a 2-element tensor.

Thanks for explaining, @apaszke! What causes confusion is that most functions treat a `tuple` or a `torch.Size` argument the same (not by design but simply due to duck typing). If function could take either a sequence or a `torch.Size` arg, there is no way to prevent this kind of ambiguity I guess.
